User Changes: Cards and Modals for Users Components. Also added notes to users.

mikedivine opened the pull request:
# Description

This PR changes the look of the user components and aligns it with the new modal card look in OED. User Notes have been added to the DB. Create button to create a user, a card for each user, edit button opens a modal that allows changes to the email, password, role, and notes.  Button to delete the user and confirmation modal.

Option is added that is labeled "Users" (only if an admin). This takes the admin to the user card page.

Partly Addresses #890 - Users Components completed only

## Type of change

- [x] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [x] I have removed text in ( ) from the issue request
- [x] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement](https://openenergydashboard.github.io/developer/cla.html) and each author is listed in the Description section.

## Limitations

The admin component changes will be a separate PR.

juanjoseguva committed:
removing user creation page and replacing it with a modal, consistent with the units creation flow

juanjoseguva committed:
Removing unused user creation page route

juanjoseguva committed:
Adding the Users page in the top menu dropdown, accessible only to admins.

juanjoseguva committed:
Removing Manage Users section in the Admin preferences page

juanjoseguva committed:
Removing Create Users page button, replacing with the modal

juanjoseguva committed:
Moving the user create button above the users table. Users table still needs to be changed to cards model, like units.

juanjoseguva committed:
Corrected password mismatch error message id

hazelcsumb committed:
Updated option values to match UserRole types for easy mapping.

Nireves333 committed:
Swap table for modals for admin user panel

juanjoseguva committed:
Adding delete user confirmation modal

juanjoseguva committed:
First commit for maps update. Changed to card view and took a stab at implementing the redux framework in MapViewComponent.

Nireves333 committed:
Implement modal view to edit User. Partial functionality, not properly setup with backend.

mikedivine committed:
uodated some files to match code organization from meters, conversions, etc. Moved files inside users folder, removed old files since we are now using modals.

Still needs the Save User Edits addressed as it's currently throwing an error.

Nireves333 committed:
Add prop for localusers

mikedivine committed:
minor code readability changes, realized api was not written to edit password and/or change the email.  Definitely would be good to rewrite api to allow password change.  Not sure if we would want to allow email change. Might be better to make them delete the account and create a new one.

mikedivine committed:
updated user api to allow for password updates

mikedivine committed:
disabled email field on edit user

Nireves333 committed:
Modify the API and slight changes to front end. Email still not being updated to DB.

Nireves333 committed:
fix merge conflict

mikedivine committed:
Updated admin user component to allow for editing user emails

mikedivine committed:
fixed admin issue, as long as current user is the only admin and not trying to switch their own role to non-admin

mikedivine committed:
update to allow users table to use notes

mikedivine committed:
updates to fix admin creation on database intilialize

mikedivine committed:
Tracked down errors for notes and got notes working! Eureka! Got rid of localUser props in user components since API no longer requires all users to be sent.

mikedivine committed:
reversing map branch changes that shouldn't be in this branch

mikedivine committed:
undo expose Postgres Server

mikedivine committed:
more map change removal

mikedivine committed:
more map reversals

mikedivine committed:
reversing, this is just a script

mikedivine committed:
again

mikedivine committed:
again

mikedivine committed:
notification format fixes, translation fixes

mikedivine committed:
Added message to notify you cannot delete your own account.

mikedivine committed:
no reason to navigate to users, already there

mikedivine committed:
No need for navigate

mikedivine committed:
changed user state to be encapsulated as userDetails

mikedivine committed:
Merge branch 'development' of https://github.com/OpenEnergyDashboard/OED into User-Changes

mikedivine committed:
merge conflicts resolved

mikedivine committed:
added MPL 2.0 headers

mikedivine committed:
didn't need this file, new version is in users folder

mikedivine committed:
updated userTest file

Nireves333 committed:
Null id fixes in userTests

huss commented on src/client/app/components/admin/users/CreateUserModalComponent.tsx:
```diff
@@ -0,0 +1,201 @@
+								</FormGroup>
+							</Col>
+						</Row>
+						{!userDetails.passwordMatch && (
```
I'm unclear on what this is for/doing. Below is FormFeedback which is used on other pages to give form feedback. This seems to try to also do that. When I commented it out it did not seem to make a difference. Should it be removed?

mikedivine replied:
Yes, now removed

huss commented on src/client/app/components/admin/users/CreateUserModalComponent.tsx:
```diff
@@ -0,0 +1,201 @@
+										type="password"
+										value={userDetails.confirmPassword}
+										onChange={e => handleStringChange(e)}
+										invalid={userDetails.confirmPassword !== userDetails.password && userDetails.confirmPassword !== ''}
```
I'm wondering if OED should warn and/or require a password to have a minimum length of say 8. I think at least doing a warning might be good. What do others think?

mikedivine replied:
I agree that this would be good. I would thinking requiring would be good along with implementing more stringent password requirements in the future would be good in my opinion. I will work on requiring the password be at least 8 chars long.

mikedivine replied:
I have implemented this change

huss commented on src/client/app/components/admin/users/CreateUserModalComponent.tsx:
```diff
@@ -0,0 +1,201 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
This is just a general note that many of the comments on this file apply to the edit user modal too.

huss commented on src/client/app/components/admin/users/CreateUserModalComponent.tsx:
```diff
@@ -0,0 +1,201 @@
+						<Row>
+							<Col>
+								<FormGroup>
+									<Label for="email">Email</Label>
```
First, this and all the labels need to be translated.

Second, I've been thinking about the use of email here, even before this PR. OED may have originally had other plans but it does not send email to users. (Sites are not required to set up email service on their servers.) I also tried and had no issues in giving a non-email input and then using it. Given this, I'm thinking it would be better to change the title and any messages that users see to "user name". What do others think? It might be nice to change all the internal names away from email but I think it is okay to leave them for now if people want to skip this step (not requires DB migration too).

mikedivine replied:
Working on changing over to username instead of email.

mikedivine replied:
Completed changing to username and all internal from email to username.

huss commented on src/client/app/components/admin/users/CreateUserModalComponent.tsx:
```diff
@@ -0,0 +1,201 @@
+										type="password"
+										value={userDetails.confirmPassword}
+										onChange={e => handleStringChange(e)}
+										invalid={userDetails.confirmPassword !== userDetails.password && userDetails.confirmPassword !== ''}
```
I'm going to suggest that the check at the end for an empty password be removed. The reason is that if no confirmation password is entered but the rest of the form is valid then the save is disabled but no error messages are shown. I do see that it leads to an immediate error after entering the password but I found that okay.

Also, the way the invalid is done here differs from edit. I wonder if it should be standardized.

mikedivine replied:
Sounds good, I will remove the check at the end for an empty password.  Also I will make the invalid password check on editUser to mirror this method

huss commented on src/client/app/components/admin/users/CreateUserModalComponent.tsx:
```diff
@@ -0,0 +1,201 @@
+	// Handlers for each type of input change
+	const handleStringChange = (e: React.ChangeEvent<HTMLInputElement>) => {
+		setUserDetails(prevDetails => ({
+			...prevDetails, ...userDetails, [e.target.name]: e.target.value
```
I'd like to understand why this needs both preDetails and userDetails? (not done in edit)

mikedivine replied:
You are correct, it does not, now removed.

huss commented on src/client/app/components/admin/users/CreateUserModalComponent.tsx:
```diff
@@ -0,0 +1,201 @@
+		email: '',
+		password: '',
+		confirmPassword: '',
+		role: '',
```
It is a little curious that the role is a string here but an enum when prepared to submit and in the form menu. I thought a little about changing this. I suspect one might need an invalid role. If it was ever sent to the server (the form should stop that) then it would be rejected by the DB because its enum for roles does not have this. Do people think it sould be changed or just leave with a comment?

mikedivine replied:
working on this. I'm planning to add an invalid role.

mikedivine replied:
I have added an invalid role.

huss commented on src/client/app/components/admin/users/CreateUserModalComponent.tsx:
```diff
@@ -0,0 +1,201 @@
+		confirmPassword: '',
+		role: '',
+		note: '',
+		passwordMatch: true
```
I have mixed feeling about having passwordMatch in the state as it can be directly calculated from other state. I might be inclined to create a function instead. However, this might just go away given a comment about its use below.

mikedivine replied:
Yes, I have decided to remove it.

mikedivine replied:
I am unsure this should go away.  I feel like it has to be checked if either password and confirmPassword changes to ensure they are matching. Otherwise if you only checked after confirmPassword, the user could go back to password and change it and there would be no check. useEffect seems the simplest method to handle this. Let me know if I am overthinking or missing something.

huss replied:
Given the added useEffect and additional checks I think it is okay to leave as is.

huss commented on src/client/app/components/admin/users/EditUserModalComponent.tsx:
```diff
@@ -0,0 +1,283 @@
+										type="password"
+										value={userDetails.confirmPassword}
+										onChange={e => handleStringChange(e)}
+										invalid={!userDetails.passwordMatch}
```
See create on comments about this and using a function instead.

mikedivine replied:
I felt checking both password fields with useEffect to see if they match seems to make the most sense to me. I am unsure of how to make it as effective another way, but I am open to your thoughts.

huss replied:
Okay per other comment.

huss commented on src/client/app/components/admin/users/CreateUserModalComponent.tsx:
```diff
@@ -0,0 +1,201 @@
+				</ModalHeader>
+				<ModalBody>
+					<Container>
+						<Row>
```
Row is defined differently in edit which, from a quick check, seems to follow other modals more (at least for unit).

mikedivine replied:
Now updated to match other modals.
I assume you meant Row xs='1' lg='2'

huss commented on src/client/app/components/admin/users/EditUserModalComponent.tsx:
```diff
@@ -0,0 +1,283 @@
+	// get current logged in user
+	const currentLoggedInUser = useAppSelector(selectCurrentUserProfile) as User;
+
+	// set user details to passed in user
```
This not only does it when first created but for any changes. Is there a case where the props change so the useEffect is needed?

mikedivine replied:
removed

mikedivine replied:
That is a good point, I don't believe props should change, so changing to just a standard useState

huss commented on src/client/app/components/admin/users/EditUserModalComponent.tsx:
```diff
@@ -0,0 +1,283 @@
+		}
+	}, [currentLoggedInUser, props.user]);
+
+	useEffect(() => {
```
The logic is not mirroring create user which does not use a useEffect. I think it would be nice for them to be parallel unless there is a reason not to.

mikedivine replied:
Understood, will have CreateUser mirror this useEffect

huss commented on src/client/app/components/admin/users/CreateUserModalComponent.tsx:
```diff
@@ -0,0 +1,201 @@
+				handleCloseModal();
+			})
+			.catch(() => {
+				showErrorNotification(translate('users.failed.to.create.user'));
```
Edit user displays the error received. Could this?

mikedivine replied:
Yes, now matches editUser

huss commented on src/client/app/components/admin/users/EditUserModalComponent.tsx:
```diff
@@ -0,0 +1,283 @@
+		// Do not call the handler function because we do not want to open the parent modal
+		setShowDeleteConfirmationModal(false);
+
+		// Delete the conversion using the state object, it should only require the source and destination ids set
```
Update comment for user not conversion state.

mikedivine replied:
done

huss commented on src/client/app/components/admin/users/EditUserModalComponent.tsx:
```diff
@@ -0,0 +1,283 @@
+						<Row>
+							<Col>
+								<FormGroup>
+									<Label for="note">Note</Label>
```
translate

mikedivine replied:
These are all using translate now.

huss commented on src/client/app/components/admin/users/CreateUserModalComponent.tsx:
```diff
@@ -0,0 +1,201 @@
+			})
+			.catch(() => {
+				showErrorNotification(translate('users.failed.to.create.user'));
+			});
```
The edit has a way to reset passwords and it does it here. I think that is a good idea for security to also do with create.

mikedivine replied:
Sounds good, will reset passwords on success or failure.

mikedivine replied:
Although, since this is a new user modal, I'll just reset all fields.

huss commented on src/client/app/components/admin/users/UserViewComponent.tsx:
```diff
@@ -0,0 +1,56 @@
+				<b><FormattedMessage id="role" />: </b>
+				{props.user.role}
+			</div>
+			<div className="edit-btn">
```
I think it would be good to show the note (limited to (I think) 30 characters) as on other cards.

mikedivine replied:
I have added the card to show the notes up to 50 chars. I tried 30, 50 looked better to me, but I can change to 30 if preferred.

huss replied:
I think it should be 30 so consistent across pages.

mikedivine replied:
Got it, changing to 30.

huss commented on src/client/app/components/admin/users/UsersDetailComponent.tsx:
```diff
@@ -0,0 +1,60 @@
+ * @returns User Detail element
+ */
+export default function UserDetailComponent() {
+	const { data: users = stableEmptyUsers } = userApi.useGetUsersQuery(undefined);
```
I removed the undefined and it worked fine. Do people think you should have that on a void method?

mikedivine replied:
Yes, it does seem to work fine without it. Now removed.

huss commented on src/client/app/components/admin/users/UsersDetailComponent.tsx:
```diff
@@ -0,0 +1,60 @@
+	const { data: users = stableEmptyUsers } = userApi.useGetUsersQuery(undefined);
+	const [localUsers, setLocalUsers] = React.useState<User[]>([]);
+
+	React.useEffect(() => { setLocalUsers(users); }, [users]);
```
Can you help me understand why there is a copy of the users state in this component?

mikedivine replied:
You are right, should not be. Removed

huss commented on src/client/app/translations/data.ts:
```diff
@@ -979,15 +984,16 @@ const LocaleTranslationData = {
+		"users.successfully.create.user": "L'utilisateur a été créé avec succès: ",
+		"users.successfully.delete.user": "L'utilisateur a été supprimé avec succès: ",
+		"users.successfully.edit.user": "L'utilisateur a été modifié avec succès: ",
+		"uses":"uses\u{26A1}",
```
Please put space back

mikedivine replied:
done.

huss commented on src/client/app/translations/data.ts:
```diff
@@ -1475,13 +1483,16 @@ const LocaleTranslationData = {
-		"users.successfully.edit.users": "Usuarios editados con éxito.",
-		"uses":"usos",
+		"user.delete.confirm": "Borrar usuario: ",
+		"user.confirm.delete": "Supprimer l'utilisateur: ",
```
This seems to be similar to the one above and not used anywhere.

mikedivine replied:
removed

huss commented on src/client/app/translations/data.ts:
```diff
@@ -1157,7 +1164,8 @@ const LocaleTranslationData = {
 		"edit.unit": "Editar unidad",
-		"email": "Correo electrónico",
+		"edit.user": "Editar Usuario",
+		"email": "Correo Electronico",
```
The change to email translation should be put back to how it was since I think that was correct.

mikedivine replied:
Email is gone now :|

huss commented on src/server/models/User.js:
```diff
@@ -12,12 +12,14 @@ class User {
+	 * @param note The user note
 	 */
-	constructor(id, email, passwordHash, role) {
+	constructor(id, email, passwordHash, role, note) {
```
There are several test files with "new User" that are missing the note. I think a clean way to deal with this addition is to have ``note = ''`` so it will default to an empty string.

mikedivine replied:
Sounds good, now note is defaulted to an empty string

huss commented on src/server/models/User.js:
```diff
@@ -49,40 +51,69 @@ class User {
 	 */
-	static async updateUserPassword(email, passwordHash, conn) {
-		return conn.none(sqlFile('user/update_user_password.sql'), { email: email, password_hash: passwordHash });
+	static async updateUserPassword(id, passwordHash, conn) {
```
src/server/services/user/editUser.js has ``await User.updateUserPassword(email, passwordHash, conn);`` that is still using the user email. It doesn't look hard to use the id from the user DB request above this.

mikedivine replied:
Ah, The editUser.js code was indented incorrectly, so I never updated that code except to fix the indentation issue. I'll look at this script and update it.

mikedivine replied:
This is now updated to use ID.

huss commented on src/server/routes/users.js:
```diff
@@ -105,22 +105,25 @@ router.post('/create', adminAuthMiddleware('create a user.'), async (req, res) =
 			role: {
 				type: 'string',
 				enum: Object.values(User.role)
+			},
```
Should note be a required param?

mikedivine replied:
I didn't make it required as I figured you could have no note, but I'll add it as required since it seems we are going to be using en ampty string as the default now.

huss commented on src/server/routes/users.js:
```diff
@@ -135,47 +138,98 @@ router.post('/create', adminAuthMiddleware('create a user.'), async (req, res) =
-
+			user: {
+				type: 'object',
+				required: ['id', 'email', 'role'],
```
require note?

mikedivine replied:
Sounds good, now required.

huss commented on src/server/routes/users.js:
```diff
@@ -135,47 +138,98 @@ router.post('/create', adminAuthMiddleware('create a user.'), async (req, res) =
 	if (!validate(req.body, validParams).valid) {
 		res.status(400).json({ message: 'Invalid params' });
 	} else {
+
```
OED does prefer not to have blank lines in places such as this.

mikedivine replied:
Got it.

huss commented on src/server/routes/users.js:
```diff
@@ -135,47 +138,98 @@ router.post('/create', adminAuthMiddleware('create a user.'), async (req, res) =
+			const numberOfAdmins = (await User.getNumberOfAdmins(conn)).count;
+			
+			// This protects the database so that there will always be at least one admin
+			if (numberOfAdmins < 2 && userBeforeChanges.role === 'admin' && user.role !== 'admin') {
```
Not a big deal but the database calls and check for # admims could be moved inside this if so it only happens if a role change is being made.

mikedivine replied:
Sounds good. I have now moved these calls inside the if statement

huss commented on src/server/routes/users.js:
```diff
@@ -135,47 +138,98 @@ router.post('/create', adminAuthMiddleware('create a user.'), async (req, res) =
+			// set up Asynchronous database queries
+			const userUpdates = [];
+
+			// not the only admin, so proceed
```
I'm okay with separating out every change but OED normally does an entire update so it is easier and one call. Now, here you cannot do the password if not provided with a new one so that could be a separate call. Up to you if you want to change this.

mikedivine replied:
Got it, it is now combined.

huss commented on src/server/services/user/createUser.js:
```diff
@@ -37,7 +37,7 @@ const { getConnection } = require('../../db');
 
 	const passwordHash = bcrypt.hashSync(password, 10);
-	const admin = new User(undefined, email, passwordHash, User.role.ADMIN);
+	const admin = new User(undefined, email, passwordHash, User.role.ADMIN, undefined);
```
In maps, it saves the calibration time with ``modifiedDate: moment().toISOString()``. OED also puts a message about automatically created units, etc. in the note. How about we put: `` `User ${email} created via createUser on ${moment().toISOString()}` `` as the note.

mikedivine replied:
Sounds good, this now has `User ${email} created via createUser on ${moment().toISOString()}` as the note.

huss commented on src/server/sql/user/insert_new_user.sql:
```diff
@@ -3,6 +3,6 @@
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 DO $$ BEGIN
```
OED does not usually do a "DO/END" in these files unless it is a function or handles an exception. Is there a special reason it is here?

mikedivine replied:
Not sure, this was here before I touched the code.  I will remove.

--------------------------------------------------------------------
Steven Huss-Lederman, 4 years ago via PR #507   (October 6th, 2020 9:17 PM) 

make createUser more resilient, fix install bug

This tried to make createUser closer to the new createDB.
The SQL avoids an error if the user exists.
createUser now stops on errors since none should happen.
A new, optional parameer to terminateReadline allows code to supply
an exit status value where default is old value of 0.
The install script for createDB retries fixes error where it stopped on
last attempt even if it worked. It also stops if createUser fails.

huss committed:
Minor updates to PR:

- Formatting
- translation order
- DB migration for updated user table
- stronger comment about functions that expose encrypted password

huss reviewed (changes requested):
Thanks to @mikedivine, @Nireves333, @hazelcsumb & @juangutierrez for this significant change. I've made some comments to address. Please let me know if anything is not clear.

I pushed a commit with small changes (see commit message) that were easy for me to do. Let me know if you think any have an issue.

mikedivine committed:
Updated backend user api and the editUser script

mikedivine committed:
updated user to use username instead of email throughout entire app, plus a few minor other user component fixes

mikedivine committed:
fixed usersTest.js - needed now required notes

mikedivine committed:
fixed some translates and other small ui tweaks

mikedivine committed:
trivial update

mikedivine committed:
altered user to only accept passwords at least 8 chars in length, fixed tooltips, fixed form validation, changed user cards to align center on new rows

mikedivine committed:
Merge branch 'OpenEnergyDashboard:development' into User-Changes

huss commented on src/client/app/components/admin/users/CreateUserModalComponent.tsx:
```diff
@@ -0,0 +1,245 @@
+	};
+
+	// allow enter key to submit form
+	const handleKeyDown = (event: React.KeyboardEvent<HTMLDivElement>) => {
```
First, this ability is only on the user pages and not any others. If we are going to keep this then it needs an issue for all pages to do this. Having said that, I have mixed feeling about having enter automatically submit the form. This and other forms have multiple buttons at the bottom so this means OED is picking submit as the default. I could see cancel as an option too. So, I'm leaning toward no default action on enter even though that can be faster. I think it is better to err on not accidentally making changes than make it easier/faster. I have a vague recollection that this may have been discussed when the modals were done and this was the decision.

Second, if this does stay then the console.log needs to be a popup and translated.

Nireves333 replied:
I'll take care of it

Nireves333 replied:
Completed. Removed the enter key functionality.

huss commented on src/server/sql/user/create_users_table.sql:
```diff
@@ -4,7 +4,8 @@
+  username VARCHAR(254) UNIQUE,
   password_hash CHAR(60) NOT NULL,
-  role user_type NOT NULL
+  role user_type NOT NULL,
```
This change needs a migration. Let me know if you need help with this.

huss commented on src/client/app/types/items.ts:
```diff
@@ -104,3 +105,15 @@ export enum TrueFalseType {
+	password: '',
+	confirmPassword: '',
+	note: '',
+	role: UserRole['Select Role'],
```
This should be set up so it is translated when shown to the user.

Nireves333 replied:
What do you mean by this?

Nireves333 replied:
We made this a default value enum, how would we go about doing this?

mikedivine replied:
I think I figured out a way and I'll implement it.

huss commented on src/client/app/translations/data.ts:
```diff
@@ -1390,7 +1399,7 @@ const LocaleTranslationData = {
 		"restore": "Restaurar",
 		"return.dashboard": "Regresar al panel principal",
-		"role": "Rol",
+		"role": "Role: ",
```
I don't think this is the Spanish word for this. Maybe add the untranslated symbol.

Nireves333 replied:
I'll fix this

Nireves333 replied:
Completed

huss commented on src/client/app/components/admin/users/EditUserModalComponent.tsx:
```diff
@@ -0,0 +1,315 @@
+										id="password"
+										name="password"
+										type="password"
+										placeholder="Only enter password to update password"
```
This string should be translated.

mikedivine replied:
Will do, I will add in translations/data.ts

huss commented on src/client/app/components/admin/users/CreateUserModalComponent.tsx:
```diff
@@ -0,0 +1,245 @@
+						<Row xs='1' lg='2'>
+							<Col>
+								<FormGroup>
+									<Label for="username">
```
I'm not sure why it passed the checks but all of these should be single quotes. I searched and we are not consistent about this but it is the OED way.

Nireves333 replied:
Will do.

Nireves333 replied:
Completed

huss commented on src/scripts/installOED.sh:
```diff
@@ -145,7 +145,7 @@ else
 	set -e
 	if [ "$production" == "no" ] && [ ! "$OED_PRODUCTION" == "yes" ]; then
-		npm run createUser -- test@example.com password
+		npm run createUser -- testUser password
```
First, given you had to edit this in two places it probably should have been a variable from the beginning. Second, while I like having shorter username without the special symbols I have mixed feeling about changing it. A lot of people expect the old email and this may throw them off. I'm inclined to either leave it alone or insert a second admin user with the new name. What do you thinkg?

mikedivine replied:
Up to you, I just thought since we were moving away from email, it would be better to use a username. I am good with whatever you prefer. I'll change it back and leave it how it was for now. Let me know and I can have it add both if you'd prefer.

huss replied:
Lets put in user called test with the same password to start to transition. Then leave a TODO to remove the old one in the future.

mikedivine replied:
Will do.

huss commented on src/server/routes/login.js:
```diff
@@ -15,16 +15,16 @@ const router = express.Router();
-			email: {
+			username: {
 				type: 'string',
 				minLength: 3,
```
Do you have any thoughts on the minLength of 3 for both username and password? This does not match the OED requirement for creation of either. I'm inclined to say it is required and put no limit here since it will mismatch if wrong. Thoughts?

Also, should we add maxLength on the password to stop bad actors or is it not really needed?

Nireves333 replied:
Having a minimum and maximum length could be a good idea for security reasons. 
I think matching them here and in the creation modal would be a good idea.

mikedivine replied:
I will remove the minLength as requested as your explanation makes sense to me.  I will add 128 as max password, to prevent bad actors mucking with it.

mikedivine replied:
let me know if you think 128 is not large enough.

huss replied:
I think we should put back minLength but set it to 8 on the password. Username can be any min length.

mikedivine replied:
OK I will add back minLength to 8 on password and 3 on username.

huss commented on src/server/routes/obvius.js:
```diff
@@ -112,17 +112,17 @@ function obviusLog(req, res, next){
 		return;
-	} else if (!req.param('email')){
-		failure(req, res, 'email parameter is required.');
+	} else if (!req.param('username')){
```
This could break scripts that send Obvius data to OED. I don't recall (and did not quickly find) if Obvius sends the email key. We need to check this out to figure out what is allowed. We also need to decide if we should allow either for backward compatibility.

mikedivine replied:
Yeah, it most likely does as I found the function obviusUsernameAndPasswordAuthMiddleware in authenticator.js and it has this:
const user = await verifyCredentials(req.body.username, req.body.password, true);

We can discuss or let me know if you prefer to allow for backward compatibility.

mikedivine replied:
altered to be backwards compatible, left note.

huss commented on src/server/test/web/obvius/README.md:
````diff
@@ -5,7 +5,7 @@ Here are some sample files for testing Obvius routes.
 ```bash
 curl localhost:3000/api/obvius -X POST \
-	-F 'email=test@example.com' \
+	-F 'username=testUser' \
````
See another comment on this change.

mikedivine replied:
Changed back for now

huss commented on src/server/test/web/csvPipelineTest.js:
```diff
@@ -21,8 +21,8 @@ const UPLOAD_METERS_ROUTE = '/api/csv/meters';
 // The default fields and routes for the requests
-const CHAI_READINGS_REQUEST = `chai.request(app).post('${UPLOAD_READINGS_ROUTE}').field('email', '${testUser.email}').field('password', '${testUser.password}')`;
-const CHAI_METERS_REQUEST = `chai.request(app).post('${UPLOAD_METERS_ROUTE}').field('email', '${testUser.email}').field('password', '${testUser.password}')`;
+const CHAI_READINGS_REQUEST = `chai.request(app).post('${UPLOAD_READINGS_ROUTE}').field('username', '${testUser.username}').field('password', '${testUser.password}')`;
```
All these changes to username for test indicate a breaking change. We need to discuss if this is okay. Right now I'm included to either put it back or accept the old and new way. Lets discuss before making any changes.

mikedivine replied:
Ok, sounds good, we can discuss at the meeting.

mikedivine replied:
I updated the obvius meter login with backwards compatibility with email and left a note. Will need documentation changes for curl commands with CSV uploads to use username instead of email moving forward.

huss commented on src/client/app/components/admin/users/UserViewComponent.tsx:
```diff
@@ -0,0 +1,64 @@
+				<b>
+					{translate('note')}
+				</b>
+				{props.user.note ? (props.user.note.length > 50 ? props.user.note.substring(0, 50) + '...' : props.user.note) : ''}
```
Frist, other pages just do the slice so maybe do that here too (if don't add the extra ...). Second, if going to add ... then it needs to be on all pages and we need to verify that 30 characters still stays on one line on all cards.

mikedivine replied:
Got it, I'll use slice and not include the '...'

huss commented on src/client/app/components/admin/users/CreateUserModalComponent.tsx:
```diff
@@ -0,0 +1,245 @@
+
+	// check if form is valid
+	const isFormValid = () => {
+		return !!userDetails.username &&
```
Should this use the state variables rather than check again? Similar comment for edit.

mikedivine replied:
Got it, now using the state variables that check on useEffect

huss commented on src/server/sql/user/create_users_table.sql:
```diff
@@ -4,7 +4,8 @@
 CREATE TABLE IF NOT EXISTS users(
   id SERIAL PRIMARY KEY NOT NULL,
-  email VARCHAR(254) UNIQUE,
+  username VARCHAR(254) UNIQUE,
```
This change needs a migration. Let me know if you need help with this.

mikedivine replied:
Ok I added the changes, I beleive just renaming the email column to username should be all that is needed.  Let me know if I missed something.

huss reviewed (changes requested):
Thanks once again to @mikedivine for addressing all the previous comments. I esp. appreciate the large number of changes due to email -> username. I think this is very close where I have some comments to think about/get input.

mikedivine committed:
fixed note to use slice and not allow note to be undefined since server will now set as empty string.

Nireves333 committed:
Double quotes fixed to single, fix translation, remove enter key create

Nireves333 committed:
merge fix

mikedivine committed:
fixed css bug

mikedivine committed:
Added db migration for users email column to username, also added a translation

mikedivine committed:
change username back to test@example.com

mikedivine committed:
altered server login min/max lengths

mikedivine committed:
changed username back to test@example.com in readme

mikedivine committed:
removed !! from isFormValid

mikedivine committed:
included currentUserSlice to include note

mikedivine committed:
using state checks from useEffect

mikedivine committed:
updated some translation settings to user components

mikedivine committed:
fixed password reset on editUser component

mikedivine committed:
Merge branch 'development' into User-Changes

mikedivine committed:
fixed userDefaults issue

mikedivine committed:
fixed obvius backwards compatibility for using email and set up tests to test for both

mikedivine committed:
Merge branch 'development' into User-Changes

mikedivine committed:
updated meter readings and csv readings/meters curl to allow email as well as username for backwards compatibility

huss commented on src/scripts/installOED.sh:
```diff
@@ -160,7 +160,7 @@ else
 		else
 			# There was no createdb error so assume database ready for use so stop process
-			printf "%s\n" "User creation had no errors so default user 'test@example.com' with password 'password' should exist"
+			printf "%s\n" "User creation had no errors so default user 'testUser' with password 'password' should exist"
```
This outputs testUser and not the actual name. I think it would be good to create a variable at the top of the script and use it here and in the npm command above.

mikedivine replied:
Will do.

huss commented on src/server/test/web/obvius/README.md:
````diff
@@ -5,7 +5,7 @@ Here are some sample files for testing Obvius routes.
 ```bash
 curl localhost:3000/api/obvius -X POST \
-	-F 'email=test@example.com' \
+	-F 'username=test@example.com' \
````
Please change this to <obvius user> instead of test@example.com (both places). This indicates it needs to be replaced and can be any user with obvius role (including admin).

mikedivine replied:
OK, I changed to test on both spots.

huss commented on src/client/app/components/admin/users/CreateUserModalComponent.tsx:
```diff
@@ -0,0 +1,234 @@
+	// check if form is valid
+	const isFormValid = () => {
+		return userDetails.username &&
+        userDetails.passwordMatch &&
```
The spacing should be tabs in a couple of places.

mikedivine replied:
Got it, will resolve this

huss commented on src/client/app/components/admin/users/UsersDetailComponent.tsx:
```diff
@@ -0,0 +1,60 @@
+							[...users]
+								.sort((a, b) => a.username.localeCompare(b.username))
+								.map(user => (
+									<Col key={user.username} className="d-flex justify-content-center mb-3">
```
Could you help me understand what the key is for?

mikedivine replied:
React requires key on root element inside the array, I'll remove from UserViewComponent since it isn't needed there.

huss commented on src/client/app/components/admin/users/CreateUserModalComponent.tsx:
```diff
@@ -0,0 +1,234 @@
+										onChange={handleRoleChange}
+										invalid={userDetails.role === UserRole['user.role.select']}
+									>
+										{Object.entries(UserRole).map(([role, value]) => (
```
This is done differently than the meter modal where the need to select is gone once you make a selection (cannot choose it again). I think that is a little nicer. Please either fix of put in a TODO.

mikedivine replied:
OK, this now works in a similar manner.

mikedivine committed:
set script to create user: test as well as test@example.com

huss commented on src/server/routes/obvius.js:
```diff
@@ -112,17 +112,23 @@ function obviusLog(req, res, next){
-	if (!req.param('password')) {
+	// First we ensure that the password and username parameters are provided.
+	const password = req.param('password');
+	// allowing for backwards compatibility if previous eGuage meters are using the 'email' parameter instead of
```
First, indenting is off. Second, could you please add a TODO marker at start of line so it will stand out more and work with TODO systems.

mikedivine replied:
Yes, will do.

huss commented on src/server/services/csvPipeline/validateCsvUploadParams.js:
```diff
@@ -92,12 +92,16 @@ const DEFAULTS = {
 const COMMON_PROPERTIES = {
 	meterName: new StringParam('meterName', undefined, undefined),
+	username: new StringParam('username', undefined, undefined),
+	// allowing for backwards compatibility to allow for curl users to use the 'email' parameter instead of
```
First, add TODO per other comment. Also, what happens if someone passes both the username and email? For example, which one is used?

mikedivine replied:
This takes the username param if it exists, otherwise it takes the email param. It only fails if neither is supplied. I tested with both in using curl and it works as intended.

huss commented on src/server/routes/obvius.js:
```diff
@@ -112,17 +112,23 @@ function obviusLog(req, res, next){
+	// allowing for backwards compatibility if previous eGuage meters are using the 'email' parameter instead of
+		// the 'username' parameter to login. Developers need to decide in the future if we should deprecate email
+		// or continue to allow this backwards compatibility
+	const username = req.param('username') || req.param('email');
```
As in another file, does this fail if they supply both? We need to have a consistent, determined result that we can doucment.

mikedivine replied:
This takes the username param if it exists, otherwise it takes the email param. It only fails if neither is supplied. I tested with both in using curl and it works as intended.

mikedivine committed:
Added minLength back in to username and password on server login

huss commented on src/server/routes/obvius.js:
```diff
@@ -112,17 +112,23 @@ function obviusLog(req, res, next){
-		obviusEmailAndPasswordAuthMiddleware('Obvius pipeline')(req, res, next);
+		req.body.username = username;
+		req.body.password = password;
+		obviusUsernameAndPasswordAuthMiddleware('Obvius pipeline')(req, res, next);
```
I'm not really knowledgeable here but I want to verify you did not break this code. The original set req.body.password and that is what src/server/routes/authenticator.js uses. Is the fact it is in req.param('password') the same? We have to be very careful with any authentication changes.

mikedivine replied:
I do not believe so. This: req.body.username = req.param('username'); I believe is equivalent to this: const username = req.param('username') || req.param('email'); req.body.username = username;

mikedivine replied:
I see some tests in the readme.md but not sure how to run them to test if everything is working properly

huss commented on src/server/models/eGauge/eGaugeRequestor.js:
```diff
@@ -19,7 +19,11 @@ class EgaugeRequestor {
 		const url = new URL(`https://${meter.url}`);
 		this.apiUrl = `${url.origin}/api`; // can we add an ending slash??
-		this.username = url.searchParams.get('username');
+		// allowing for backwards compatibility if previous eGuage meters are using the 'email' parameter instead of
```
A couple of thoughts. First, TODO per other comments. Second, issue if provide both to settle.

mikedivine replied:
works the same as the others, takes username if it exists, otherwise takes email. If neither, then fails

huss commented on src/server/test/web/csvPipelineTest.js:
```diff
@@ -472,6 +476,13 @@ const testCases = {
 		responseCode: [200, 200],
 		responseString: ['<h1>SUCCESS</h1>Successfully inserted the meters.', '<h1>SUCCESS</h1><h2>It looks like the insert of the readings was a success.</h2><h3>However, note that the processing of the readings returned these warning(s):</h3><br>For meter pipe123: Warning parsing Reading #2. Reading value gives 30 with warning message:<br>The reading start time is shifted and within the DST shift so  ...
+	},
+	pipe1: {
```
A test key should never be reused. It is a unique identifier. This should be a new test number - maybe pipe130? The same applies for the other test changed.

mikedivine replied:
Got it. I will change these.

huss commented on src/server/routes/csv.js:
```diff
@@ -67,9 +67,13 @@ router.use(function (req, res, next) {
-					const verifiedUser = await verifyCredentials(email, password, true);
+					// username and password to be supplied for curl requests.
+					const { username, email, password } = request.body;
+					// allowing for backwards compatibility if previous eGuage meters are using the 'email' parameter instead of
```
Same thoughts as other places it was changed.

mikedivine replied:
Same as others, added TODO, username parameter is taken if it is given, otherise email parameter is taken as username. If both are supplied, username takes priority. If niether is given, then fails.

huss reviewed (changes requested):
Thanks to the team for the updates. This is looking good overall and works as expected. I've made a few comments to address including on some previous ones. I'm hoping the changes are not very much.

mikedivine committed:
fixed my mistake on OED install script

mikedivine committed:
updated create user script to allow for usernames with 4 chars, updated install script to use more descriptive variable names as requested

mikedivine committed:
Create User Modal: altered userRole input selecion to remove the initial invalid value after a valid role is chosen from the drop down

mikedivine committed:
Added TODOs and fixed some tests

huss committed:
consistently enforce 3 character on user name

- Also limits route lengths to avoid issues.

huss committed:
Merge remote-tracking branch 'origin/development' into pr/mikedivine/1306-5

huss commented:
I just pushed two items via commits:

1. Added code to more consistently enforce min/max on user info.
2. Merged development to remove conflict.

I did this because I accidentally merged the new admin settings page so I wanted to quickly get this merged so people have access to the user page. If anyone sees any issue then let me know.

huss commented:
The new mins on user name means that if someone had one of less than 3 it would not longer work. Since they used to be emails this seems safe.

Though unlikely, a now invalid password could be present. If this is the case then it would need to be reset.

huss reviewed (approved):
Thanks to @mikedivine for these last changes. I appreciate your willingness to consider changes and for making the all even when that mean new items. I've tested the code and reviewed. It all seems fine.
